source/common/upstream/cluster_manager_impl.cc

```diff
@@ -1277,15 +1277,17 @@ ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::~ClusterEntry()
 
 Http::ConnectionPool::Instance*
 ClusterManagerImpl::ThreadLocalClusterManagerImpl::ClusterEntry::connPool(
-    ResourcePriority priority, Http::Protocol protocol, LoadBalancerContext* context) {
+    ResourcePriority priority, Upstream::ClusterManager::ProtocolResolutionFunc protocol,
+    LoadBalancerContext* context) {
   HostConstSharedPtr host = lb_->chooseHost(context);
   if (!host) {
     ENVOY_LOG(debug, "no healthy host for HTTP connection pool");
     cluster_info_->stats().upstream_cx_none_healthy_.inc();
     return nullptr;
   }
 
-  std::vector<uint8_t> hash_key = {uint8_t(protocol)};
+  auto final_protocol = protocol(*host);
+  std::vector<uint8_t> hash_key = {uint8_t(final_protocol)};
```

Comment:
Could we just do host.cluster().upstreamHttpProtocol(protocol); here and drop the argument?

Reply:
We need to somehow provide the downstream protocol to `upstreamHttpProtocol`, which we don't always have, e.g. when doing tcp_proxy tunneling. 

I think the main alternative would be to pass the downstream protocol as `absl::optional<Http::Protocol`, but I'm not sure if that's much better as it makes the CM have to care about what's going on on the downstream side.

Reply:
oh yeah,  I glazed over the lambda capture somewhere, sorry about that.

I think I'd prefer to pass in the optional - the function in question
ClusterInfoImpl::upstreamHttpProtocol() literally takes in downstream_protocol so and I think it's easier to read without the function obscuring what's going on (which is that we're selecting upstream protocol based on downstream protocol)